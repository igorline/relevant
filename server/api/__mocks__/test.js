// exports.test = 'test';

module.exports = 'test';
