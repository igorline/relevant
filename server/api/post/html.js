import request from 'request-promise-any';
import jsdom from 'jsdom';
import { Readability } from 'readability/index';
import cheerio from 'cheerio';
import { parse as parseUrl } from 'url';

/* eslint no-console: 0 */

const fbHeader = {
  'User-Agent':
    'facebookexternalhit/1.1 (+http://www.facebook.com/externalhit_uatext.php)'
};

const relevantHeader = {
  'User-Agent':
    'This is a user request for article matadata for Relevant Community (https://relevant.community | info@relevant.community)'
};

const URL_REGEX = new RegExp(
  /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%_+~#=]{2,256}\.[a-z]{2,10}\b([-a-zA-Z0-9@:%_+~#?&//=]*)/
);
const HTML_REGEX = new RegExp(/<[^>]*>/, 'gm');

exports.extractDomain = url => {
  let domain;
  if (url.indexOf('://') > -1) {
    domain = url.split('/')[2];
  } else {
    domain = url.split('/')[0];
  }
  domain = domain.split(':')[0];

  let noPrefix = domain;

  if (domain.indexOf('www.') > -1) {
    noPrefix = domain.replace('www.', '');
  }
  return noPrefix;
};

function stripHTML(text) {
  return (text || '').replace(HTML_REGEX, '');
}

exports.getReadable = async uri => {
  let article;
  try {
    const body = await request({
      url: uri,
      jar: true,
      headers:
        uri.match('bloomberg.com') || uri.match('washingtonpost.com')
          ? relevantHeader
          : fbHeader
    });

    // let doc = new JSDOMParser().parse(body);

    const doc = jsdom.jsdom(body, {
      features: {
        FetchExternalResources: false,
        ProcessExternalResources: false
      }
    });
    article = new Readability(uri, doc).parse();
    if (!article) throw new Error("Couldn't parse doc");
    // console.log(article.byline);
  } catch (err) {
    console.log('ERROR getting readable version', uri);
    throw err;
  }
  return article;
};

exports.trimToLength = (doc, length) => {
  Array.prototype.slice
    .call(doc.getElementsByTagName('figure'))
    .forEach(item => item.remove());

  let totalLength = 0;
  Array.prototype.slice.call(doc.getElementsByTagName('*')).forEach(el => {
    if (totalLength > length) return el.remove();
    const elChildNode = el.childNodes;
    elChildNode.forEach(child => {
      if (child.nodeType === 3) {
        const text = child.textContent;
        const l = text.split(/\s+/).length;
        if (l > 4) totalLength += l;
      }
    });
    return null;
  });
  return doc;
};

exports.generatePreview = async (body, uri, reqUrl, noReadability) => {
  // console.log('Generate Preview ', uri);

  body = body.replace('<!--', '').replace('-->', '');
  let $ = cheerio.load(body);

  const redirect = $("meta[http-equiv='refresh']")[0];
  let redirectUrl;

  if (redirect && redirect.attribs && redirect.attribs.content) {
    redirectUrl = redirect.attribs.content.split('URL=')[1];
    if (redirectUrl) redirectUrl = redirectUrl.replace(/'/g, '');
    if (
      redirectUrl &&
      uri.match('twitter.com') &&
      redirectUrl.match('mobile.twitter.com')
    ) {
      redirectUrl = uri;
    }
  }

  if (
    redirectUrl &&
    redirectUrl.match('http') &&
    exports.extractDomain(redirectUrl) !== '' &&
    exports.extractDomain(redirectUrl) !== exports.extractDomain(uri) &&
    reqUrl !== redirectUrl
  ) {
    return {
      redirect: true,
      uri: redirectUrl
    };
  }

  let canonical = $("link[rel='canonical']")[0];
  if (canonical && canonical.attribs) canonical = canonical.attribs;
  else canonical = null;

  if (canonical && canonical.href) {
    if (!URL_REGEX.test(canonical.href)) {
      canonical.href = exports.extractDomain(uri) + canonical.href;
    }
  }

  if (
    canonical &&
    canonical.href &&
    canonical.href.match('http') &&
    exports.extractDomain(canonical.href) !== exports.extractDomain(uri) &&
    reqUrl !== canonical.href
  ) {
    return {
      redirect: true,
      uri: canonical.href
    };
  }

  if (uri.match('apple.news')) {
    try {
      const articleUrl = body
        .match(/redirectToUrl\("(.*)"/)[0]
        .replace(/redirectToUrl\(|"/g, '');
      if (articleUrl) {
        return {
          redirect: true,
          uri: articleUrl
        };
      }
    } catch (err) {
      console.log(err);
    }
  }

  let data = {
    title: null,
    description: null,
    image: null,
    'og:type': null,
    'og:title': null,
    'og:description': null,
    'og:image': null,
    'og:image:url': null,
    'og:url': null,
    'al:web:url': null,
    'twitter:title': null,
    'twitter:image': null,
    'twitter:image:src': null,
    'twitter:description': null,
    'twitter:site': null,
    'twitter:creator': null,
    news_keywords: null,
    keywords: null,
    'article:tag': null,
    author: null,
    'article:author': null
  };

  // ads milti element concotenation
  const miltiElement = [
    'news_keywords',
    'keywords',
    'article:tag',
    'author',
    'article:author'
  ];
  const meta = $('meta');
  const $title = $('title');
  const keys = Object.keys(meta);

  data.title = $title.eq(0).text();

  Object.keys(data).forEach(s => {
    keys.forEach(key => {
      if (
        meta[key].attribs &&
        (meta[key].attribs.property === s ||
          meta[key].attribs.name === s ||
          meta[key].attribs.itemprop === s)
      ) {
        if (!data[s]) data[s] = meta[key].attribs.content;
        else if (miltiElement.find(m => m === s)) {
          data[s] += ', ' + meta[key].attribs.content;
        }
      }
    });
  });

  let description = null;
  let title = null;
  let image = null;

  title = data['og:title'] || data['twitter:title'] || data.title;
  description = stripHTML(
    data.description || data['og:description'] || data['twitter:description']
  ).trim();
  image =
    data['og:image'] ||
    data['og:image:url'] ||
    data['twitter:image'] ||
    data['twitter:image:src'] ||
    data.image;

  // why prioritise og tags? flipboard?
  let url;
  if (uri.match('flip.it')) {
    url = data['al:web:url'] || data['og:url'] || uri;
  }
  url = uri || data['al:web:url'] || data['og:url'];
  const tags = data.news_keywords || data.keywords;
  const domain = exports.extractDomain(url);

  const originalUrl = parseUrl(uri);
  const cannonicalUrl = parseUrl(url);

  const k1 = data.keywords ? data.keywords.split(',').map(k => k.trim()) : [];
  const k2 = data.news_keywords ? data.news_keywords.split(',').map(k => k.trim()) : [];
  const k3 = data['article:tag']
    ? data['article:tag']
        .split(',')
        .map(k => k.replace('--primarykeyword-', '').trim())
        .filter(k => !k.match('--'))
    : [];

  let keywords = [...new Set([...k1, ...k2, ...k3])];

  // general fix for mismatch
  if (originalUrl.pathname !== cannonicalUrl.pathname) {
    url = originalUrl.protocol + '//' + originalUrl.host + originalUrl.pathname;
  }

  let amp = $('[type="application/ld+json"]')
    .eq(0)
    .text();
  let ampKeywords;
  let ampAuthor;
  if (amp) {
    try {
      amp = amp
        .replace('//<![CDATA[', '')
        .replace('// <![CDATA[', '')
        .replace('//]]>', '')
        .replace('// ]]>', '')
        .trim();
      amp = JSON.parse(amp);
      if (!amp) throw new Error('no amp');
      ampAuthor = amp.author ? amp.author.name : null;
      ampKeywords = amp.keywords && amp.keywords.length ? amp.keywords : null;
      if (typeof ampAuthor === 'string') ampAuthor = [ampAuthor];
    } catch (err) {
      // console.log(amp);
      // console.log(err);
    }
  }
  // console.log('amp: ', amp ? JSON.parse(amp) : null);

  // console.log('url ', url);
  // console.log('article:author ', data['article:author']);
  // console.log('author ', data['author']);
  // console.log('amp author ', ampAuthor);
  // console.log('regular keywords ', keywords);
  // console.log('amp keywords', ampKeywords);

  data.author = data.author || '';

  data['article:author'] = data['article:author'] || '';

  const metaAuthor = data.author
    .split(/,|\sand\s/)
    .map(a => a.trim())
    .filter(a => a !== '');

  ampAuthor = ampAuthor || [];

  let author = [...ampAuthor, ...metaAuthor];
  author = [...new Set(author)].filter(a => !a.match('http'));

  if (ampKeywords && ampKeywords.length && !keywords.length) {
    if (Array.isArray(ampKeywords)) {
      ampKeywords = ampKeywords.filter(k => !k.match('@')).map(k => k.toLowerCase());
    }
  }

  keywords = keywords.map(k => k.toLowerCase());
  keywords = [...new Set([...(keywords || []), ...(ampKeywords || [])])];

  keywords = keywords
    .map(k => k.replace(/tag:|publication:|elevated:false|lockedpostsource:0/, ''))
    .filter(k => k.trim() !== '');

  let doc = jsdom.jsdom(body, {
    features: {
      FetchExternalResources: false,
      ProcessExternalResources: false
    }
  });

  let article;

  if (!noReadability) {
    try {
      article = new Readability(url, doc).parse();
    } catch (err) {
      console.log('Readability ERR ', err);
    }
  }

  let short;
  if (article && article.article) {
    try {
      short = article.excerpt || exports.trimToLength(article.article, 140).innerHTML;
    } catch (err) {
      console.log(err);
    }
    if (!author.length && article.byline) {
      const by = article.byline
        .split(/\n|,|â€¢/)
        .map(a => a.replace(/by|by:\s/i, '').trim())
        .filter(
          a =>
            a !== '' &&
            !a.match(/^by$/i) &&
            a.length > 2 &&
            !a.match('UTC') &&
            Number.isNaN(new Date(a).getTime()) &&
            !a.match('http')
          // && !a.match(/2017|2016|2015|2018/)
        );
      author = by[0] ? [by[0]] : [];
    }
  }

  if (image && image.indexOf('http://') !== 0 && image.indexOf('https://') !== 0) {
    image = 'http://' + domain + image;
  }

  if (!title && url.match('.pdf')) {
    title = url.substring(url.lastIndexOf('/') + 1);
  }

  const obj = {
    image,
    description,
    title,
    url,
    domain,
    tags,
    shortText: short,
    articleAuthor: author || null,
    keywords
  };

  if (!image || !title) {
    // console.log('url parse incomplete');
    // console.log(data);
    // console.log(uri);
    // console.log($('head').html());
  }

  $ = null;
  body = null;
  article = null;
  data = null;
  doc = null;

  return {
    redirect: false,
    result: obj
  };
};
