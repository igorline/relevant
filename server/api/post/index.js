const express = require('express');
const controller = require('./post.controller');
const auth = require('../../auth/auth.service');
const { asyncMiddleware } = require('../../utils/middlewares');

const router = express.Router();

router.post('/', auth.isAuthenticated(), auth.communityMember(), controller.create);
router.post(
  '/sendPostNotification',
  auth.hasRole('admin'),
  controller.sendPostNotification
);

router.put('/', auth.isAuthenticated(), auth.communityMember(), controller.update);
router.put('/flag', auth.isAuthenticated(), controller.flag);

router.get('/flagged', auth.hasRole('admin'), controller.flagged);
// router.get('/', auth.authMiddleware(), controller.index);
router.get('/readable', controller.readable);
router.get('/topPosts', controller.topPosts);
router.get('/:id', auth.blocked(), asyncMiddleware(controller.index));
router.get('/user/:id', auth.blocked(), controller.userPosts);
router.get('/preview/generate', controller.preview);

router.delete('/:id', auth.isAuthenticated(), controller.remove);

module.exports = router;
