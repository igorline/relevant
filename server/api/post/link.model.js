const mongoose = require('mongoose');

const { Schema } = mongoose;

// TODO check for unused fields
const MetaPostSchema = new Schema(
  {
    title: String,
    description: String,
    image: String,
    url: { type: String, index: true },

    shortText: { type: String },
    longText: { type: String },
    articleDate: Date,
    articleAuthor: [String],
    copyright: String,
    links: [
      {
        text: String,
        href: String
      }
    ],
    publisher: String,
    domain: String,
    keywords: [String],

    tags: [{ type: String, ref: 'Tag' }],
    // DEPRECATED?
    categories: [{ type: String, ref: 'Tag' }],

    flagged: { type: Boolean, default: false },
    flaggedBy: [{ type: String, ref: 'User', select: false }],
    flaggedTime: Date
  },
  {
    timestamps: true
  }
);

MetaPostSchema.index({ url: 1 });
MetaPostSchema.index({ rank: 1 });
MetaPostSchema.index({ latestPost: 1 });
MetaPostSchema.index({ latestPost: 1, tags: 1 });
MetaPostSchema.index({ rank: 1, tags: 1 });

module.exports = mongoose.model('MetaPost', MetaPostSchema);
