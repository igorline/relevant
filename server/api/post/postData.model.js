const mongoose = require('mongoose');

const { Schema } = mongoose;

// TODO USE THIS
const PostDataSchema = new Schema(
  {
    post: { type: Schema.Types.ObjectId, ref: 'Post' },
    community: String,
    communityId: { type: Schema.Types.ObjectId, ref: 'Community' },

    eligibleForReward: { type: Boolean, default: false },

    rank: { type: Number, default: 0 },
    relevance: { type: Number, default: 0 },
    pagerank: { type: Number, default: 0 },
    pagerankRaw: { type: Number, default: 0 },
    pagerankRawNeg: { type: Number, default: 0 },

    relevanceNeg: { type: Number, default: 0 },

    // should we track community comments separately?
    // commentCount: { type: Number, default: 0 },
    upVotes: { type: Number, default: 0 },
    downVotes: { type: Number, default: 0 },

    postDate: { type: Date, index: true },
    latestComment: { type: Date, index: true },

    paidOut: { type: Boolean, default: false },
    payoutTime: { type: Date },
    payout: { type: Number, default: 0 },
    payOutShare: { type: Number, default: 0 },
    expectedPayout: { type: Number, default: 0 },

    shares: { type: Number, default: 0 },
    balance: { type: Number, default: 0 },
    totalShares: { type: Number, default: 0 }, // track positive and negative here

    latestTweet: { type: Date },

    // NEW
    isInFeed: { type: Boolean, default: false },
    repost: { type: Boolean, default: false },
    type: { type: String, default: 'post' },
    parentPost: { type: Schema.Types.ObjectId, ref: 'Post' },
    parentComment: { type: Schema.Types.ObjectId, ref: 'Post' },
    hidden: { type: Boolean, default: false },
    tags: [String],
    needsRankUpdate: Boolean
  },
  {
    timestamps: true
  }
);

PostDataSchema.index({ post: 1 });
PostDataSchema.index({ post: 1, community: 1 });
PostDataSchema.index({ post: 1, communityId: 1 });
PostDataSchema.index({ isInFeed: 1 });

PostDataSchema.index({ latestComment: -1, community: 1 });
PostDataSchema.index({ isInFeed: 1, community: 1, latestComment: -1 });
PostDataSchema.index({ isInFeed: 1, community: 1, rank: -1 });

PostDataSchema.index({ isInFeed: 1, communityId: 1, rank: -1 });
PostDataSchema.index({ isInFeed: 1, communityId: 1, latestComment: -1 });

PostDataSchema.index({ post: 1, communityId: 1 });
PostDataSchema.index({ parentPost: 1, communityId: 1 });
PostDataSchema.index({ isInFeed: 1, communityId: 1, rank: 1 });

module.exports = mongoose.model('PostData', PostDataSchema);
