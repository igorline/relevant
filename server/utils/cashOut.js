import Earnings from '../api/earnings/earnings.model';

export async function logCashOut(user, cashOutAmt, next) {
  try {
    const earnings = await Earnings.find({
      user: user._id,
      status: 'paidout',
      earned: { $gt: 0 }
    });
    const totalPreviousPaidout = earnings.reduce(getTotalEarnings, 0);
    const cashOutLog = new Earnings({
      user: user._id,
      cashOutAttempt: true,
      cashOutAmt,
      prevBalance: user.balance,
      endBalance: user.balance - cashOutAmt,
      totalPreviousPaidout,
      legacyAirdrop: user.legacyAirdrop,
      referralTokens: user.referralTokens,
      airdropTokens: user.airdropTokens,
      status: 'in progress',
      source: 'wallet'
    });
    await cashOutLog.save();
    return cashOutLog;
  } catch (err) {
    return next(new Error('Error logging cashOut', err));
  }
}

export function getTotalEarnings(total, currentVal) {
  total += currentVal.earned;
  return total;
}
