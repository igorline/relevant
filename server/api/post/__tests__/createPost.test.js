import Post from 'server/api/post/post.model';
import PostData from 'server/api/post/postData.model';
import Link from 'server/api/post/link.model';
import CommunityMember from 'server/api/community/community.member.model';
import { response } from 'jest-mock-express';
import { create, remove } from 'server/api/post/post.controller';
import { post, community } from 'app/mockdata';
import { sanitize } from 'server/test/utils';
import { getUsers } from 'server/test/seedData';

// jest.mock('request-promise-native');

let alice;
const { createPost } = post;
const { relevant } = community;

// this will define the database name where the tests are run
process.env.TEST_SUITE = 'createPost';

describe('CreatePost', () => {
  let linkPostId;
  let linkParentId;
  let res;
  let req;
  const next = console.log; // eslint-disable-line

  beforeEach(() => {
    res = response();
    ({ alice } = getUsers());
    req = {
      user: alice,
      body: createPost,
      communityMember: {
        community: relevant.slug,
        communityId: relevant._id
      }
    };
  });

  describe('api/post/create', () => {
    test('create link post', async () => {
      await create(req, res, next);
      let apiRes = res.json.mock.calls[0][0].toObject();
      expect(apiRes.parentPost).toBeTruthy();
      expect(apiRes.parentPost.data.isInFeed).toBe(true);
      expect(apiRes.data.isInFeed).toBe(false);
      expect(apiRes.image).toBe(createPost.image);
      expect(apiRes.data.eligibleForReward).toBe(false);
      linkPostId = apiRes._id;
      linkParentId = apiRes.parentPost._id;
      apiRes = sanitize(apiRes, 'post');
      expect(apiRes).toMatchSnapshot();
    });

    test('expect parentPost to be created', async () => {
      const postParent = await Post.findOne({ _id: linkParentId }).populate({
        path: 'data',
        match: { communityId: relevant._id }
      });
      expect(postParent.data.eligibleForReward).toBe(true);
      expect(postParent.title).toBe(createPost.title);
      expect(postParent.image).toBe(createPost.image);
      expect(postParent.url).toBe(createPost.url);
      expect(postParent.description).toBe(createPost.description);
    });

    test('create text post', async () => {
      const newReq = { ...req };
      delete newReq.body.url;
      await create(newReq, res, next);
      let apiRes = res.json.mock.calls[0][0].toObject();
      expect(apiRes.parentPost).not.toBeDefined();
      expect(apiRes.data.isInFeed).toBe(true);
      apiRes = sanitize(apiRes, 'post');
      expect(apiRes).toMatchSnapshot();
    });

    test('community member unreads should update', async () => {
      const member = await CommunityMember.findOne({
        communityId: relevant._id,
        user: alice._id
      });
      expect(member.unread).toBeGreaterThan(0);
    });
  });

  describe('api/post/delete', () => {
    test('delete link post', async () => {
      const delReq = {
        user: alice,
        params: { id: linkPostId }
      };
      await remove(delReq, res, next);
      const apiRes = res.json.mock.calls[0][0];
      expect(apiRes).toEqual('removed');
    });
  });

  describe('After remove', () => {
    test('should remove parent', async () => {
      const doc = await Post.findOne({ _id: linkParentId });
      expect(doc).toBeNull();
    });
    test('should remove parent data', async () => {
      const doc = await PostData.findOne({ post: linkParentId });
      expect(doc).toBeNull();
    });
    test('should remove data', async () => {
      const doc = await PostData.findOne({ post: linkPostId });
      expect(doc).toBeNull();
    });
    test('should remove metaPost', async () => {
      const doc = await Link.findOne({ post: linkPostId });
      expect(doc).toBeNull();
    });
  });
});
