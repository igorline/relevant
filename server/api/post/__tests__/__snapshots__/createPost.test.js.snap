// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePost api/post/create create link post 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "body": "awesome post #1!",
  "category": null,
  "channel": false,
  "commentCount": 0,
  "community": "relevant",
  "communityId": "c00000000000000000000001",
  "createdAt": "[SANITIZED]",
  "data": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "balance": 0,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "downVotes": 0,
    "eligibleForReward": false,
    "expectedPayout": 0,
    "hidden": false,
    "isInFeed": false,
    "latestComment": "[SANITIZED]",
    "pagerank": 0,
    "pagerankRaw": 0,
    "pagerankRawNeg": 0,
    "paidOut": false,
    "parentPost": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "channel": false,
      "commentCount": 1,
      "createdAt": "[SANITIZED]",
      "data": Object {
        "__v": 0,
        "_id": "[SANITIZED]",
        "balance": 0,
        "community": "relevant",
        "communityId": "c00000000000000000000001",
        "createdAt": "[SANITIZED]",
        "downVotes": 0,
        "eligibleForReward": true,
        "expectedPayout": 0,
        "hidden": false,
        "isInFeed": true,
        "latestComment": "[SANITIZED]",
        "pagerank": 0,
        "pagerankRaw": 0,
        "pagerankRawNeg": 0,
        "paidOut": false,
        "payOutShare": 0,
        "payout": 0,
        "payoutTime": "[SANITIZED]",
        "post": "[SANITIZED]",
        "postDate": "[SANITIZED]",
        "rank": "[SANITIZED]",
        "relevance": 0,
        "relevanceNeg": 0,
        "repost": false,
        "shares": 0,
        "tags": Array [
          "tag11",
          "tag12",
        ],
        "totalShares": 0,
        "type": "link",
        "upVotes": 0,
        "updatedAt": "[SANITIZED]",
      },
      "description": "my amazing post",
      "downVotes": 0,
      "flagged": false,
      "flaggedBy": Array [],
      "hidden": false,
      "id": "[SANITIZED]",
      "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
      "latestComment": "[SANITIZED]",
      "links": Array [],
      "mentions": Array [],
      "metaPost": Object {
        "__v": 0,
        "_id": "[SANITIZED]",
        "articleAuthor": Array [],
        "categories": Array [
          null,
        ],
        "createdAt": "[SANITIZED]",
        "description": "my amazing post",
        "domain": "example.com",
        "flagged": false,
        "flaggedBy": Array [],
        "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
        "keywords": Array [],
        "links": Array [],
        "tags": Array [
          "tag11",
          "tag12",
        ],
        "title": "postTitle",
        "updatedAt": "[SANITIZED]",
        "url": "https://example.com/testPost1",
      },
      "pagerank": 0,
      "paidOut": false,
      "payoutTime": "[SANITIZED]",
      "postDate": "[SANITIZED]",
      "rank": 0,
      "relevance": 0,
      "seenInFeedNumber": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "title": "postTitle",
      "twitter": false,
      "twitterScore": 0,
      "type": "link",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
      "url": "https://example.com/testPost1",
      "version": "metaRework",
    },
    "payOutShare": 0,
    "payout": 0,
    "payoutTime": "[SANITIZED]",
    "post": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "relevanceNeg": 0,
    "repost": false,
    "shares": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "totalShares": 0,
    "type": "post",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
  },
  "description": null,
  "downVotes": 0,
  "embeddedUser": Object {
    "_id": "[SANITIZED]",
    "handle": "alice",
    "image": "https://hips.hearstapps.com/elleuk.cdnds.net/15/37/2048x2730/2048x2730-e6-8278-458c32fb7c74-assets-elleuk-com-gallery-17063-taylor-swift-hairstyle-bob-getty-gallery-01-jpg.jpg",
    "name": "Alice",
  },
  "flagged": false,
  "flaggedBy": Array [],
  "hidden": false,
  "id": "[SANITIZED]",
  "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
  "linkParent": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "channel": false,
    "commentCount": 1,
    "createdAt": "[SANITIZED]",
    "data": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "balance": 0,
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "downVotes": 0,
      "eligibleForReward": true,
      "expectedPayout": 0,
      "hidden": false,
      "isInFeed": true,
      "latestComment": "[SANITIZED]",
      "pagerank": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "paidOut": false,
      "payOutShare": 0,
      "payout": 0,
      "payoutTime": "[SANITIZED]",
      "post": "[SANITIZED]",
      "postDate": "[SANITIZED]",
      "rank": "[SANITIZED]",
      "relevance": 0,
      "relevanceNeg": 0,
      "repost": false,
      "shares": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "totalShares": 0,
      "type": "link",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
    },
    "description": "my amazing post",
    "downVotes": 0,
    "flagged": false,
    "flaggedBy": Array [],
    "hidden": false,
    "id": "[SANITIZED]",
    "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
    "latestComment": "[SANITIZED]",
    "links": Array [],
    "mentions": Array [],
    "metaPost": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "articleAuthor": Array [],
      "categories": Array [
        null,
      ],
      "createdAt": "[SANITIZED]",
      "description": "my amazing post",
      "domain": "example.com",
      "flagged": false,
      "flaggedBy": Array [],
      "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
      "keywords": Array [],
      "links": Array [],
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "title": "postTitle",
      "updatedAt": "[SANITIZED]",
      "url": "https://example.com/testPost1",
    },
    "pagerank": 0,
    "paidOut": false,
    "payoutTime": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "seenInFeedNumber": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "twitter": false,
    "twitterScore": 0,
    "type": "link",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
    "version": "metaRework",
  },
  "links": Array [],
  "mentions": Array [
    "test",
    "user1",
  ],
  "metaPost": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "articleAuthor": Array [],
    "categories": Array [
      null,
    ],
    "createdAt": "[SANITIZED]",
    "description": "my amazing post",
    "domain": "example.com",
    "flagged": false,
    "flaggedBy": Array [],
    "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
    "keywords": Array [],
    "links": Array [],
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
  },
  "pagerank": 0,
  "paidOut": false,
  "parentPost": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "channel": false,
    "commentCount": 1,
    "createdAt": "[SANITIZED]",
    "data": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "balance": 0,
      "community": "relevant",
      "communityId": "c00000000000000000000001",
      "createdAt": "[SANITIZED]",
      "downVotes": 0,
      "eligibleForReward": true,
      "expectedPayout": 0,
      "hidden": false,
      "isInFeed": true,
      "latestComment": "[SANITIZED]",
      "pagerank": 0,
      "pagerankRaw": 0,
      "pagerankRawNeg": 0,
      "paidOut": false,
      "payOutShare": 0,
      "payout": 0,
      "payoutTime": "[SANITIZED]",
      "post": "[SANITIZED]",
      "postDate": "[SANITIZED]",
      "rank": "[SANITIZED]",
      "relevance": 0,
      "relevanceNeg": 0,
      "repost": false,
      "shares": 0,
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "totalShares": 0,
      "type": "link",
      "upVotes": 0,
      "updatedAt": "[SANITIZED]",
    },
    "description": "my amazing post",
    "downVotes": 0,
    "flagged": false,
    "flaggedBy": Array [],
    "hidden": false,
    "id": "[SANITIZED]",
    "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
    "latestComment": "[SANITIZED]",
    "links": Array [],
    "mentions": Array [],
    "metaPost": Object {
      "__v": 0,
      "_id": "[SANITIZED]",
      "articleAuthor": Array [],
      "categories": Array [
        null,
      ],
      "createdAt": "[SANITIZED]",
      "description": "my amazing post",
      "domain": "example.com",
      "flagged": false,
      "flaggedBy": Array [],
      "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
      "keywords": Array [],
      "links": Array [],
      "tags": Array [
        "tag11",
        "tag12",
      ],
      "title": "postTitle",
      "updatedAt": "[SANITIZED]",
      "url": "https://example.com/testPost1",
    },
    "pagerank": 0,
    "paidOut": false,
    "payoutTime": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "seenInFeedNumber": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "title": "postTitle",
    "twitter": false,
    "twitterScore": 0,
    "type": "link",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
    "url": "https://example.com/testPost1",
    "version": "metaRework",
  },
  "payoutTime": "[SANITIZED]",
  "postDate": "[SANITIZED]",
  "rank": 0,
  "relevance": 0,
  "seenInFeedNumber": 0,
  "tags": Array [
    "tag11",
    "tag12",
  ],
  "title": "postTitle",
  "twitter": false,
  "twitterScore": 0,
  "type": "post",
  "upVotes": 0,
  "updatedAt": "[SANITIZED]",
  "url": "https://example.com/testPost1",
  "user": "a00000000000000000000000",
  "version": "metaRework",
}
`;

exports[`CreatePost api/post/create create text post 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "body": "awesome post #1!",
  "category": null,
  "channel": false,
  "commentCount": 0,
  "community": "relevant",
  "communityId": "c00000000000000000000001",
  "createdAt": "[SANITIZED]",
  "data": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "balance": 0,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "downVotes": 0,
    "eligibleForReward": true,
    "expectedPayout": 0,
    "hidden": false,
    "isInFeed": true,
    "latestComment": "[SANITIZED]",
    "pagerank": 0,
    "pagerankRaw": 0,
    "pagerankRawNeg": 0,
    "paidOut": false,
    "payOutShare": 0,
    "payout": 0,
    "payoutTime": "[SANITIZED]",
    "post": "[SANITIZED]",
    "postDate": "[SANITIZED]",
    "rank": 0,
    "relevance": 0,
    "relevanceNeg": 0,
    "repost": false,
    "shares": 0,
    "tags": Array [
      "tag11",
      "tag12",
    ],
    "totalShares": 0,
    "type": "post",
    "upVotes": 0,
    "updatedAt": "[SANITIZED]",
  },
  "description": null,
  "downVotes": 0,
  "embeddedUser": Object {
    "_id": "[SANITIZED]",
    "handle": "alice",
    "image": "https://hips.hearstapps.com/elleuk.cdnds.net/15/37/2048x2730/2048x2730-e6-8278-458c32fb7c74-assets-elleuk-com-gallery-17063-taylor-swift-hairstyle-bob-getty-gallery-01-jpg.jpg",
    "name": "Alice",
  },
  "flagged": false,
  "flaggedBy": Array [],
  "hidden": false,
  "id": "[SANITIZED]",
  "image": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/cute-cat-captions-1563551865.jpg",
  "links": Array [],
  "mentions": Array [
    "test",
    "user1",
  ],
  "pagerank": 0,
  "paidOut": false,
  "payoutTime": "[SANITIZED]",
  "postDate": "[SANITIZED]",
  "rank": 0,
  "relevance": 0,
  "seenInFeedNumber": 0,
  "tags": Array [
    "tag11",
    "tag12",
  ],
  "title": "postTitle",
  "twitter": false,
  "twitterScore": 0,
  "type": "post",
  "upVotes": 0,
  "updatedAt": "[SANITIZED]",
  "user": "a00000000000000000000000",
  "version": "metaRework",
}
`;
